### PR TITLE
Concurrent rabbit dequeue

### PR DESCRIPTION
**Pull request by mbaechler**
With this PR we remove entirely the Spooler threadpool and replace it with a Reactor concurrency level tuning parameter (it has been tested locally and has the desired effect).

### CODE REVIEW

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java [hunk 1]
```diff
@@ -101,47 +92,38 @@ public void configure(HierarchicalConfiguration<ImmutableNode> config) {
...
-                .doOnSuccess(any -> timeMetric.stopAndPublish())
-                .doOnSuccess(any -> processingActive.decrementAndGet());
-        } catch (Throwable e) {
-            return Mono.error(e);
-        }
+        return metricFactory
+            .runPublishingTimerMetric(SPOOL_PROCESSING,
+                 Mono.fromCallable(processingActive::incrementAndGet)
+                    .flatMap(ignore -> processMail(queueItem))
+                    .doOnSuccess(any -> processingActive.decrementAndGet()));
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: Not related, but this is not decremented if processing fails?

File: server/queue/queue-api/src/main/java/org/apache/james/queue/api/MailQueueFactory.java [hunk 2]
```diff
@@ -38,9 +63,17 @@
      * @param name
      * @return queue
      */
-    Optional<T> getQueue(MailQueueName name);
+    default Optional<T> getQueue(MailQueueName name) {
+        return getQueue(name, defaultPrefetchCount());
+    }
+
+    Optional<T> getQueue(MailQueueName name, PrefetchCount count);
```
Comment (chibenwa) [inline server/queue/queue-api/src/main/java/org/apache/james/queue/api/MailQueueFactory.java]: Can we get javadoc on what prefetch count is?

File: server/queue/queue-api/src/main/java/org/apache/james/queue/api/MailQueueFactory.java [hunk 3]
```diff
@@ -27,6 +27,31 @@
...
+
+    static PrefetchCount prefetchCount(int count) {
+        return new PrefetchCount(count);
+    }
+
+    class PrefetchCount {
+        private final int value;
+
+        private PrefetchCount(int value) {
+            this.value = value;
```
Comment (chibenwa) [inline server/queue/queue-api/src/main/java/org/apache/james/queue/api/MailQueueFactory.java]: Can prefetch count be negative?

**Comment by Arsnael** [comment]
```
[5217700e8d10c9c01d2b51d60b42d305a6d32e48] [ERROR] Failed to execute goal on project apache-james-backends-jpa: Could not resolve dependencies for project org.apache.james:apache-james-backends-jpa:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre: Failed to read artifact descriptor for com.google.guava:guava:jar:25.1-jre: Could not transfer artifact com.google.guava:guava:pom:25.1-jre from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/com/google/guava/guava/25.1-jre/guava-25.1-jre.pom.part (No such file or directory) -> [Help 1]
```
test this please

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java [hunk 4]
```diff
@@ -101,47 +92,38 @@ public void configure(HierarchicalConfiguration<ImmutableNode> config) {
     @PostConstruct
     public void init() {
         LOGGER.info("init...");
-        queue = queueFactory.createQueue(MailQueueFactory.SPOOL);
-        spooler = Schedulers.fromExecutor(Executors.newFixedThreadPool(numThreads, NamedThreadFactory.withName("spooler")));
-        LOGGER.info("uses {} Thread(s)", numThreads);
-        run();
+        LOGGER.info("Concurrency level is {}", concurrencyLevel);
+        queue = queueFactory.createQueue(MailQueueFactory.SPOOL, MailQueueFactory.prefetchCount(concurrencyLevel));
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: I bet you set the prefetch count equal to the concurrency level as rabbitMQ overhead is considered neglectible compared with the mail processing time.

Correct? 

If not, can you provide rationals for this choice?

Comment (mbaechler) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: PrefetchCount, for the case of Rabbit will translate to "how many elements can be not-acked-yet at the same time". Any number less than your expected concurrency would make the concurrency less than asked. If you choose a number greater than you concurrency, you will have a less fair distribution of mail dequeues in your cluster.

At first, I was planning something like `concurrency * 1.1` but it doesn't work for small number. Then `concurrency * 1.1 + 1` but I then fell "why should I care for such micro details upfront?" In worst case, we'll have a small pipelining latency. Does it worth the added complexity of having a not-so-clear math formula?

However, I can change my mind you think we should invest in that.

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java (hunk 1 above)
Comment (mbaechler) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: yes, good catch, will fix

File: server/queue/queue-api/src/main/java/org/apache/james/queue/api/MailQueueFactory.java (hunk 3 above)
Comment (mbaechler) [reply server/queue/queue-api/src/main/java/org/apache/james/queue/api/MailQueueFactory.java]: I guess no (:

File: server/queue/queue-api/src/main/java/org/apache/james/queue/api/MailQueueFactory.java (hunk 2 above)
Comment (mbaechler) [reply server/queue/queue-api/src/main/java/org/apache/james/queue/api/MailQueueFactory.java]: Yes

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java (hunk 4 above)
Comment (chibenwa) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: No that's perfect, that's a great explanation.

However isn't it a rabbitMQ specific detail somehow being hardcoded in a generic part of the code?

I understand the needs, I just wonder if we have some cleaner ways of linking the implementation to the usage? Maybe with a guice injection?

Comment (mbaechler) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: As we use reactive types, it makes sense to let the previous stages of a Publisher/Flux to know how much prefetch we need as it's defined for unitary operations like flatMap, etc.

For now we don't use it on other mailqueues but we probably could if load testing shows specific problems on another implementation we care about.

**Comment by nvduc91** [comment]
Read it

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java (hunk 4 above)
Comment (chibenwa) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: I was thinking: How about using a `ConcurrencyLevel` object in the queue API, then we can have this "prefect = concurrencyLevel" rule in rabbitMQMailQueue, along with a nice comment explaining it? WDYT?

Comment (mbaechler) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: I don't get your suggestion. Is it just renaming PrefetchCount type to ConcurrencyLevel? Is it something else?

Comment (chibenwa) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: > Is it just renaming PrefetchCount type to ConcurrencyLevel?

Yes.

That allow relocating the actual decision of "the prefetch is the concurrency level" in rabbit backend, and add a comment explaining why.

Sorry I may be picky, but trying to keep things generic enough in APIs seems important to me.

Comment (mbaechler) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: A `Flux` doesn't have a concurrency level, you need a `ParallelFlux` to express that.

For now, we only use `Publisher` type and it's closer to `Flux` than `ParallelFlux` so I would not be comfortable to define a concurrency property.

For `Flux` concurrency doesn't exist, only operators have that concept.

However `Flux` does have a `prefetch` concept (how many elements are flowing at a given time).

So i'm picky too but I prefer prefetch. Let me know if you disagree on some of my arguments.

Comment (chibenwa) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: We're not speaking of the concurrency level of the resulting flux but that the underlying mailQueue is expected to handle.

Comment (mbaechler) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java]: > We're not speaking of the concurrency level of the resulting flux but that the underlying mailQueue is expected to handle.

The mailqueue doesn't care about concurrency, it's only a consumer concern.

However, how many pending elements you can take from the queue is something the queue has to manage to enable concurrency on the consumer side.

> To be noted, reactor is not even part of the MailQueue API, it's an implementation detail.

Publisher is part of MailQueue API

Screenshot from 2020-02-28 16-00-00

**Comment by Arsnael** [comment]
Log output lost

test this please

**Comment by mbaechler** [comment]
test this please

**Comment by chibenwa** [comment]
```
[f7d0d072aee5fa7971f1560557b17c3af8904797] [ERROR]   JamesMailSpoolerTest.threadSuicideShouldAckTheItem:120 » NullPointer
[f7d0d072aee5fa7971f1560557b17c3af8904797] [ERROR]   JamesMailSpoolerTest.thrownExceptionShouldAckTheItem:85 » NullPointer
```

&&

```

[f7d0d072aee5fa7971f1560557b17c3af8904797] [ERROR] thrownExceptionShouldAckTheItem  Time elapsed: 0.958 s  <<< ERROR!
[f7d0d072aee5fa7971f1560557b17c3af8904797] java.lang.NullPointerException
[f7d0d072aee5fa7971f1560557b17c3af8904797] 	at org.apache.james.mailetcontainer.impl.JamesMailSpoolerTest.thrownExceptionShouldAckTheItem(JamesMailSpoolerTest.java:85)
[f7d0d072aee5fa7971f1560557b17c3af8904797] 
[f7d0d072aee5fa7971f1560557b17c3af8904797] [ERROR] threadSuicideShouldAckTheItem  Time elapsed: 0.004 s  <<< ERROR!
[f7d0d072aee5fa7971f1560557b17c3af8904797] java.lang.NullPointerException
[f7d0d072aee5fa7971f1560557b17c3af8904797] 	at org.apache.james.mailetcontainer.impl.JamesMailSpoolerTest.threadSuicideShouldAckTheItem(JamesMailSpoolerTest.java:120)
```

Related?

test this please

**Comment by blackheaven** [comment]
* master:
master1

* this PR:
after1
after2

I run the remaining ones

**Comment by blackheaven** [comment]
No logs

test this please

**Comment by chibenwa** [comment]
```
[f7d0d072aee5fa7971f1560557b17c3af8904797] [ERROR]   JamesMailSpoolerTest.threadSuicideShouldAckTheItem:120 » NullPointer
[f7d0d072aee5fa7971f1560557b17c3af8904797] [ERROR]   JamesMailSpoolerTest.thrownExceptionShouldAckTheItem:85 » NullPointer
```

test this please

**Comment by mbaechler** [comment]
> Conclusion: equivalent, but more stable

Or more lucky

**Comment by mbaechler** [comment]
push forced to rebase against master

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java [hunk 5]
```diff
@@ -185,7 +172,7 @@ public void dispose() {
 
     @Override
     public int getThreadCount() {
```
Comment (remk) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java:170]: Should we rename the method in the parent interface MailSpoolerMBean ?

**Comment by Arsnael** [comment]
CI restart

test this please

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java (hunk 5 above)
Comment (mbaechler) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java:170]: it's part of the public API, it's why I didn't changed it for now

**Comment by mbaechler** [comment]
The load-testing we did show no regression regarding enqueuing. We should see better figures on dequeue but we lack a proper way to measure that aside configuring and looking at a grafana board.

I'm working on a metrics logger to solve that but suggest we merge this PR regardless as local tests proved it work really well.

**Comment by mbaechler** [comment]
> I'm working on a metrics logger to solve that but suggest we merge this PR regardless as local tests proved it work really well.

see https://issues.apache.org/jira/browse/JAMES-3145

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**